### ISSUE TITLE
Add "calendar" style

### ISSUE THREAD

**Issue by Turbo87** [yahoo/intl-relativeformat#33]
[Moment.js](http://momentjs.com/docs/#/displaying/calendar-time/) has a "calendar" style output, that works quite well in some situations. It would be awesome if this project would support that output style too.

**Comment by caridy** [comment]
@Turbo87 I don't see that information in CLDR, maybe @rxaviers knows if that even exists in CLDR.

**Comment by Turbo87** [comment]
@caridy I've done some research on that and it seems to exist in some form.

[ca-generic.json](https://github.com/unicode-cldr/cldr-dates-full/blob/master/main/en/ca-generic.json#L298) contains `{1} 'at' {0}`, the [dateFields.json](https://github.com/unicode-cldr/cldr-dates-full/blob/master/main/en/dateFields.json#L247-L251) contains stuff like `last Monday`, and the time is just formatted as usual.

**Comment by rxaviers** [comment]
Hi!

The data is there in CLDR ([example](https://github.com/unicode-cldr/cldr-dates-modern/blob/master/main/en/dateFields.json#L248)). Although, relative weekday formatting is tough because it's poorly specified by UTS#35 and there is enough confusion about it (moment/moment#1411).

Formatting relative weekday is tightly coupled with the use case. The reference changes depending of the use case. Here I had summarized my findings https://github.com/moment/moment/issues/1411#issuecomment-45262235. I hope it helps.

**Comment by caridy** [comment]
@rxaviers I was talking about the composition of the sentence (e.g.: the `at`... in `en`).

@Turbo87 that data is what we use in `Intl` polyfill:

```
            "dateTimeFormats": {
              "full": "{1} 'at' {0}",
              "long": "{1} 'at' {0}",
              "medium": "{1}, {0}",
              "short": "{1}, {0}",
              "availableFormats": {
                "E": "ccc",
                "EHm": "E HH:mm",
                "EHms": "E HH:mm:ss",
                "Ed": "d E",
                "Ehm": "E h:mm a",
                "Ehms": "E h:mm:ss a",
                "Gy": "y G",
                "GyMMM": "MMM y G",
                "GyMMMEd": "E, MMM d, y G",
                "GyMMMd": "MMM d, y G",
                "H": "HH",
                "Hm": "HH:mm",
                "Hms": "HH:mm:ss",
                "M": "L",
                "MEd": "E, M/d",
                "MMM": "LLL",
                "MMMEd": "E, MMM d",
                "MMMd": "MMM d",
                "Md": "M/d",
                "d": "d",
                "h": "h a",
                "hm": "h:mm a",
                "hms": "h:mm:ss a",
                "ms": "mm:ss",
                "y": "y G",
                "yyyy": "y G",
                "yyyyM": "M/y GGGGG",
                "yyyyMEd": "E, M/d/y GGGGG",
                "yyyyMMM": "MMM y G",
                "yyyyMMMEd": "E, MMM d, y G",
                "yyyyMMMd": "MMM d, y G",
                "yyyyMd": "M/d/y GGGGG",
                "yyyyQQQ": "QQQ y G",
                "yyyyQQQQ": "QQQQ y G"
              }
```

specifically, in `Intl` polyfill I use `medium` as a way to compile more complex date formats from those available in `availableFormats`, by combining those that are time specific `{0}` and those that are date specific `{1}`. It might be true that we can use `long` or `full` to construct the calendar style, but we will have more research around that to confirm that this is the case :)

**Comment by caridy** [comment]
There is also the question about when to use the extended format or not, because `Last month at 2:30 PM` makes no sense, we will need that kind of information as well.

**Comment by rxaviers** [comment]
> @rxaviers I was talking about the composition of the sentence (e.g.: the at... in en).

Oh! Yeap, the `dateTimeFormats` `full|long|medium|short` may provide the glue. On http://www.unicode.org/reports/tr35/tr35-dates.html#availableFormats_appendItems (at the bottom of this subsection, item 3), there's a combination algorithm for combining date fields with time fields. I am not sure if the same applies for relative date fields.

**Comment by Turbo87** [comment]
@caridy how did you come up with `Last month at 2:30 PM`? this example should fall in the "Everything else" category and would be formatted as a date instead. (e.g. `7/10/2011`)

**Comment by caridy** [comment]
@Turbo87 right, the question is: where is that rule defined? I don't think that exists in CLDR, we need a proper algo definition.

**Comment by Turbo87** [comment]
@caridy I would mimic the [moment.js algo](https://github.com/moment/moment/blob/develop/src/lib/moment/calendar.js#L10-L15) for now:

``` js
        format = diff < -6 ? 'sameElse' :
            diff < -1 ? 'lastWeek' :
            diff < 0 ? 'lastDay' :
            diff < 1 ? 'sameDay' :
            diff < 2 ? 'nextDay' :
            diff < 7 ? 'nextWeek' : 'sameElse';
```

**Comment by caridy** [comment]
hmm, will that work for every lang? I will like to see some formalization around this concept.
